Framework/Crystal/src/SCDCalibratePanels2.cpp

```diff
@@ -524,36 +500,351 @@ void SCDCalibratePanels2::optimizeBanks(std::shared_ptr<PeaksWorkspace> pws) {
+          }
+        }
+      }
+      niter += 1;
+      stepsSum = 0.0;
+      for (int i = 0; i < 6; ++i)
+        stepsSum += searchStep[i];
+
+      // logging
+      V3D xyz = V3D(params[0], params[1], params[2]);
+      double _theta = params[3];
+      double _phi = params[4];
+      double _ang = params[5];
+      V3D rotvec =
+          V3D(sin(_theta) * cos(_phi), sin(_theta) * sin(_phi), cos(_theta));
+      std::ostringstream logmsg;
+      logmsg.precision(12);
+      logmsg << bankname << "@iter_" << niter << "\n"
+             << "-- best_err: " << best_err << "\n"
+             << "-- (dx,dy,dz): " << xyz << "\n"
+             << "-- ang@rotvec: " << _ang << "@" << rotvec << "\n";
+      g_log.notice() << logmsg.str();
+    }
+
+    //-- step 2: move bank to corresponding location
+    const double dx = params[0];
+    const double dy = params[1];
+    const double dz = params[2];
+    const double theta = params[3];
+    const double phi = params[4];
+    const double rotang = params[5];
+    double rvx = sin(theta) * cos(phi);
+    double rvy = sin(theta) * sin(phi);
+    double rvz = cos(theta);
+    std::string bn = bankname;
+    if (pws->getInstrument()->getName().compare("CORELLI") == 0)
+      bn.append("/sixteenpack");
+    adjustComponent(dx, dy, dz, rvx, rvy, rvz, rotang, bn, pws);
+
+    //-- step 3: log
+    V3D dxyz = V3D(dx, dy, dz);
+    V3D rv = V3D(rvx, rvy, rvz);
+    std::ostringstream msg;
+    msg.precision(8);
+    msg << bankname << " calibration results:\n"
+        << "(dx, dy, dz) = " << dxyz << "\n"
+        << "ang@rot-axis = " << rotang << "@" << rv << "\n";
+    g_log.notice() << msg.str();
+
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+}
+
+double SCDCalibratePanels2::objfunc_bank(IPeaksWorkspace_sptr pwsBank,
+                                         double params[6],
+                                         std::string bankname) {
+  // prep
+  double err = 0.0;
+  V3D qv_target;
+  V3D qv_calc;
+  V3D delta_qv;
+  Units::Wavelength wl;
+
+  // only operate on the clone
+  IPeaksWorkspace_sptr pws = pwsBank->clone();
+
+  // purturb the instrument
+  const double dx = params[0];
+  const double dy = params[1];
+  const double dz = params[2];
+  const double theta = params[3];
+  const double phi = params[4];
+  const double rotang = params[5];
+  double rvx = sin(theta) * cos(phi);
+  double rvy = sin(theta) * sin(phi);
+  double rvz = cos(theta);
+  std::string bn = bankname;
+  if (pws->getInstrument()->getName().compare("CORELLI") == 0)
+    bn.append("/sixteenpack");
+  adjustComponent(dx, dy, dz, rvx, rvy, rvz, rotang, bn, pws);
+
+  // calculate the residual
+  auto ubmatrix = pws->sample().getOrientedLattice().getUB();
+  for (int i = 0; i < pws->getNumberPeaks(); ++i) {
+    qv_target = ubmatrix * pwsBank->getPeak(i).getIntHKL();
+    qv_target *= 2 * PI;
+
+    // cache time-of-flight
+    double tof = pws->getPeak(i).getTOF();
```

Comment:
I'm confused as to why you use the old detector positions to get `tof`, and then use this `tof` later to obtain the wavelength of a peak with new detector positions.

Reply:
similar reason as above.